crypto/engine/eng_devcrypto.c

```diff
@@ -587,8 +924,121 @@ static int devcrypto_digests(ENGINE *e, const EVP_MD **digest,
+#define DEVCRYPTO_CMD_CIPHERS (ENGINE_CMD_BASE + 1)
+#define DEVCRYPTO_CMD_DIGESTS (ENGINE_CMD_BASE + 2)
+#define DEVCRYPTO_CMD_DUMP_INFO (ENGINE_CMD_BASE + 3)
+
+static const ENGINE_CMD_DEFN devcrypto_cmds[] = {
+   {DEVCRYPTO_CMD_USE_SOFTDRIVERS,
+    "USE_SOFTDRIVERS",
+    "specifies whether to use software (not accelerated) drivers (" STR(DEVCRYPTO_REQUIRE_ACCELERATED)
+        "=use only accelerated drivers, " STR(DEVCRYPTO_USE_SOFTWARE) "=allow all drivers, "
+        STR(DEVCRYPTO_REJECT_SOFTWARE) "=use if acceleration can't be determined [default=2])",
+    ENGINE_CMD_FLAG_NUMERIC},
+
+   {DEVCRYPTO_CMD_CIPHERS,
+    "CIPHERS",
+    "either ALL, NONE, or a comma-separated list of ciphers to enable [default=ALL]",
+    ENGINE_CMD_FLAG_STRING},
+
+#ifdef IMPLEMENT_DIGEST
+   {DEVCRYPTO_CMD_DIGESTS,
+    "DIGESTS",
+    "either ALL, NONE, or a comma-separated list of digests to enable [default=ALL]",
+    ENGINE_CMD_FLAG_STRING},
+#endif
+
+   {DEVCRYPTO_CMD_DUMP_INFO,
+    "DUMP_INFO",
+    "dump info about each algorithm to stderr; use 'openssl engine -pre DUMP_INFO devcrypto'",
+    ENGINE_CMD_FLAG_NO_INPUT},
+
+   {0, NULL, NULL, 0}
+};
+
+static int devcrypto_ctrl(ENGINE *e, int cmd, long i, void *p, void (*f) (void))
+{
+    switch(cmd) {
+    case DEVCRYPTO_CMD_USE_SOFTDRIVERS:
+        switch(i) {
+            case DEVCRYPTO_REQUIRE_ACCELERATED:
+            case DEVCRYPTO_USE_SOFTWARE:
+            case DEVCRYPTO_REJECT_SOFTWARE:
+                break;
+            default:
+                fprintf(stderr, "devcrypto: invalid value (%ld) for USE_SOFTDRIVERS\n", i);
+            return 0;
+        }
+        if (use_softdrivers == i)
+            return 1;
+        use_softdrivers = i;
+#ifdef IMPLEMENT_DIGEST
+        rebuild_known_digest_nids();
+#endif
+        rebuild_known_cipher_nids();
+        return 1;
+
+    case DEVCRYPTO_CMD_CIPHERS:
+        if (p == NULL)
+            return 1;
+        if (strcasecmp((const char *)p, "ALL") == 0) {
+            cryptodev_enable_all_ciphers();
+            goto reload_ciphers;
+        }
+        cryptodev_disable_all_ciphers();
+        if (strcasecmp((const char*)p, "NONE") == 0)
+            goto reload_ciphers;
+        else if (!CONF_parse_list(p, ',', 1, cryptodev_enable_cipher_cb, NULL)) {
+            cryptodev_enable_all_ciphers();
+            return 0;
+        }
+reload_ciphers:
+        rebuild_known_cipher_nids();
+        return 1;
+
+#ifdef IMPLEMENT_DIGEST
+    case DEVCRYPTO_CMD_DIGESTS:
+        if (p == NULL)
+            return 1;
+        if (strcasecmp((const char *)p, "ALL") == 0) {
+            cryptodev_enable_all_digests();
+            goto reload_digests;
+        }
+        cryptodev_disable_all_digests();
+        if (strcasecmp((const char*)p, "NONE") == 0)
+            goto reload_digests;
+        else if (!CONF_parse_list(p, ',', 1, cryptodev_enable_digest_cb, NULL)) {
+            cryptodev_enable_all_digests();
```

Comment:
Same comment here...  I'm thinking that if parsing the string fails, the previous state of the digests should be restored, no?

Reply:
I've done this.  It will only get restored in case of a strndup failure, really.